src/main/java/com/datapipe/jenkins/vault/buildstep/VaultStepExecution.java

```diff
@@ -0,0 +1,107 @@
+import org.jenkinsci.plugins.workflow.steps.StepExecution;
+
+import com.cloudbees.plugins.credentials.CredentialsProvider;
+import com.datapipe.jenkins.vault.VaultAccessor;
+import com.datapipe.jenkins.vault.configuration.GlobalVaultConfiguration;
+import com.datapipe.jenkins.vault.configuration.VaultConfiguration;
+import com.datapipe.jenkins.vault.credentials.VaultCredential;
+import com.datapipe.jenkins.vault.exception.VaultPluginException;
+
+import hudson.EnvVars;
+import hudson.Util;
+import hudson.model.Run;
+import hudson.model.TaskListener;
+
+public class VaultStepExecution extends StepExecution {
+    private final transient VaultReadStep readStep;
+
+    private final EnvVars envVars;
+    private final transient Run run;
+    private final transient TaskListener listener;
+
+    VaultStepExecution(@Nonnull final StepContext context, final VaultReadStep readStep) throws IOException, InterruptedException {
+        super(context);
+        this.readStep = readStep;
+        this.run = getRunFromContext();
+        this.listener = getTaskListenerFromContext();
+        this.envVars = run.getEnvironment(listener);
+    }
+
+    @Override
+    public boolean start() {
+        try {
+
+            final String path = Util.replaceMacro(readStep.getPath(), envVars);
+            final int engineVersion = readStep.getEngineVersion();
+            final VaultAccessor vaultAccessor = initVaultAccessor();
+
+            final String value = vaultAccessor.read(path, engineVersion)
+                .getData()
+                .get(Util.replaceMacro(readStep.getKey(), envVars));
+            getContext().onSuccess(value);
+        } catch (VaultPluginException ex) {
+            getContext().onFailure(ex);
+        }
+        return true;
+    }
+
+    @Override
+    public void stop(@Nonnull final Throwable cause) {
+        // no action required on stopping
+    }
+
+    private VaultAccessor initVaultAccessor() {
+        GlobalVaultConfiguration globalVaultConfiguration = GlobalVaultConfiguration.all().get(GlobalVaultConfiguration.class);
+        if (globalVaultConfiguration == null) {
+            throw new VaultPluginException("Unable to retrieve GlobalVaultConfiguration");
+        }
+
+        final VaultConfiguration vaultConfiguration = globalVaultConfiguration.getConfiguration();
+        final String credentialsId = readStep.getCredentialsId() == null || readStep.getCredentialsId().isEmpty() ? vaultConfiguration.getVaultCredentialId() : Util.replaceMacro(readStep.getCredentialsId(), envVars);
+        final String vaultUrl = readStep.getVaultUrl() == null || readStep.getVaultUrl().isEmpty() ? vaultConfiguration.getVaultUrl() : Util.replaceMacro(readStep.getVaultUrl(), envVars);
+        final boolean skipSslVerification = vaultConfiguration.isSkipSslVerification();
+
+        VaultAccessor vaultAccessor = new VaultAccessor();
+
+        if (credentialsId == null || credentialsId.isEmpty()) {
```

Comment:
I suggest using `StringUtils.isBlank()`
```suggestion
        if (StringUtils.isBlank(credentialsId)) {
```